src/NUnitFramework/framework/Attributes/DataSourceAttribute.cs

```diff
@@ -0,0 +1,392 @@
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System;
+using System.IO;
+using System.Xml;
+using System.Collections;
+using System.Collections.Generic;
+using System.Reflection;
+using NUnit.Framework.Interfaces;
+using NUnit.Framework.Internal;
+using NUnit.Framework.Internal.Builders;
+
+namespace NUnit.Framework
+{
+
+  /// <summary>
+  /// The type of external data source
+  /// </summary>
+  public enum DataSourceType: int {
+    /// <summary>
+    /// No datas ource is specified. This is the default state.
+    /// </summary>
+    Null = 0,
+    
+    /// <summary>
+    /// Data source is a comma separated value (CSV) file.
+    /// </summary>
+    Csv = 1,
+    
+    /// <summary>
+    /// Data source is an XML document
+    /// </summary>
+    Xml = 2,
+    
+    /// <summary>
+    /// Data source is a database
+    /// </summary>
+    Db = 4
+  };
+  
+  /// <summary>
+  /// Common amounts of data rows to collect
+  /// </summary>
+  public enum DataRowCount: int {
+    /// <summary>
+    /// Read all rows
+    /// </summary>
+    All = 0
+  }
+  
+  
+  /// <summary>
+  /// Specifies an external data source for test methods. This is evaulated at compile time.
+  /// </summary>
+  [AttributeUsage(AttributeTargets.Method, AllowMultiple = true, Inherited = false)]
+  public class DataSourceAttribute : NUnitAttribute, ITestBuilder, IImplyFixture
+  {
+    private NUnitTestCaseBuilder _builder = new NUnitTestCaseBuilder();
+
+    #region Constructors
+
+    /// <summary>
+    /// Constructs the object responsible for obtaining test data.
+    /// </summary>
+    /// <param name="dataType">The type of data provider</param>
+    /// <param name="sourceName">The connection string for the data source. It could be a filename or DB connection string.</param>
+    /// <param name="rowCount">The number of rows to collect from the data source. By default it is all rows</param>
+    public DataSourceAttribute(DataSourceType dataType, string sourceName, int rowCount = (int)DataRowCount.All)
+    {
+      this.SourceName = sourceName;
+      this.DataType = dataType;
+      this.DataRows = rowCount;
+    }
+    
+    #endregion
+
+    #region Properties
+
+    /// <summary>a
+    /// The name of a file or connection string which contains the data
+    /// </summary>
+    public string SourceName { get; private set; }
+
+    /// <summary>
+    /// The data input type
+    /// </summary>
+    public DataSourceType DataType {get; private set;}
```

Comment:
I dislike XxxxType as a name pattern when it's not a System.Type.

An alternative to having this enum as an argument would be to derive three different attributes for Csv, Xml and database respectively.